## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [79]:
import pandas as pd

client_data = pd.read_csv('./Resources/client_dataset.csv')

client_data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [80]:
# View the column names in the data
column_header_data = client_data.columns
print(column_header_data)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [81]:
# Use the describe function to gather some basic statistics
basic_stats = client_data.describe(include='all')
print(basic_stats)

              first     last        job       phone                     email  \
count         54639    54639      54639       54639                     54639   
unique          360      477        498         999                       999   
top     Christopher  Johnson  Osteopath  5016915923  jessica.reyes@grimes.net   
freq           1090     1164        631         220                       220   
mean            NaN      NaN        NaN         NaN                       NaN   
std             NaN      NaN        NaN         NaN                       NaN   
min             NaN      NaN        NaN         NaN                       NaN   
25%             NaN      NaN        NaN         NaN                       NaN   
50%             NaN      NaN        NaN         NaN                       NaN   
75%             NaN      NaN        NaN         NaN                       NaN   
max             NaN      NaN        NaN         NaN                       NaN   

           client_id      o

In [82]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# shape_df = df.shape
# print(shape_df)

# size_df = df.size
# print(size_df)

# type_df = df.dtypes
# print(type_df)

info_df = client_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [83]:
# What three item categories had the most entries?
# sort_data_by_cat = df["category"].sort_values()
# reset_index_sorted_data = sort_data_by_cat.reset_index(drop=True)
value_counts_df = client_data["category"].value_counts()
# print(sort_data_by_cat)
# print(reset_index_sorted_data)
top_3_counts_df = value_counts_df.head(3)
# print(value_counts_df)
print(top_3_counts_df)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [84]:
# For the category with the most entries,
# which subcategory had the most entries?
sub_value_counts_df = client_data[["category", "subcategory"]].value_counts()
# top_sub_counts_df = sub_value_counts_df.head(1)
sub_value_counts_df.head(1)
# print(top_sub_counts_df)

category     subcategory      
consumables  bathroom supplies    6424
Name: count, dtype: int64

In [85]:
# Which five clients had the most entries in the data?
top_5_clients_df = client_data[["first", "last", "client_id"]].value_counts()
top_5_clients_df.head()

first      last     client_id
Jessica    Reyes    33615        220
Angela     Everett  66037        211
Bryan      Myers    46820        209
Kendra     Garrett  24741        207
Alexandra  Young    38378        207
Name: count, dtype: int64

In [86]:
# Store the client ids of those top 5 clients in a list.
top_clients_id_df = client_data["client_id"].value_counts()
top_5_clients_df = top_clients_id_df.head(5).index.tolist()
top_5_clients_df

[33615, 66037, 46820, 38378, 24741]

In [87]:
# How many total units (the qty column) did the
# client with the most entries order order?
client_qty_ordered_df = client_data.groupby("client_id")["qty"].sum()
is_client_id = 33615
is_client_qty = 220
client_known_data = client_qty_ordered_df.loc[is_client_id]
print("Client " + str(is_client_id) + " has ordered " + str(is_client_qty) + " times, for a total of " + str(client_known_data) + " items.")

Client 33615 has ordered 220 times, for a total of 64313 items.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [88]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
client_data["subtotal"] = client_data["unit_price"] * client_data["qty"]
client_data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [89]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
# client_data["shipping price"] = lambda unit_weight, qty: 7 * unit_weight if unit_weight * qty > 50 else 10  )
# client_data["shipping price"] = lambda x: 7 * (client_data["qty"] * client_data["unit_weight"]) if client_data["unit_weight"] * client_data["qty"] > 50 else 10 * (client_data["qty"] * client_data["unit_weight"])
# cal_shipping_cost = lambda unit_weight, qty: 7 * (qty * unit_weight) if (unit_weight * qty) > 50 else 10 * (qty * unit_weight)
# client_data['shipping_price'] = cal_shipping_cost(client_data['unit_weight'], client_data['qty'])
cal_shipping_cost = lambda x: 7 * (x['qty'] * x['unit_weight']) if (x['unit_weight'] * x['qty']) > 50 else 10 * (x['qty'] * x['unit_weight'])
client_data['shipping price'] = client_data.apply(cal_shipping_cost, axis=1)

client_data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [90]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
sales_tax = 1.0925
total_price = (client_data["subtotal"] + client_data["shipping price"]) * float(sales_tax)
client_data["total price"] = round(total_price, 2)
client_data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping price  \
0     762.71         7.50  105            1  115164.00         5512.50   
1      15.09         1.49   21            0     523.95          312.90   
2       7.86         1.68   39            6     527.28          458.64   
3      24.85         1.23   29            3    1056.18          356.70   
4     108.17        46.43   20            1    3902.00         6500.20   

   total price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 21 columns]

In [100]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
client_data["new total cost"] = (client_data["unit_cost"] * client_data["qty"]) + client_data["shipping price"]
client_data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping price  total price  new total cost  \
0            1  115164.00         5512.50    131839.08        85597.05   
1            0     523.95          312.90       914.26          629.79   
2            6     527.28          458.64      1077.12          765.18   
3            3    1056.18          356.70      1543.57         1077.35   
4            1    3902.00         6500.20     11364.40         8663.60   

     profit  
0 -46242.03  
1   -284.47  
2   -311.94  
3   -466.22  
4  -2700.80  

[5 rows x 23 columns]

In [120]:
# Create a column for the profit of
# each line using line cost and line price
client_data["profit"] = client_data["total price"] - client_data["new total cost"]
client_data

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  shipping price  total price  new total cost  \
0                1  115164.00         5512.50    131839.08        85597.05   
1                0     523.95          312.90       914.26          629.79   
2                6     527.28          458.64      1077.12          765.18   
3                3    1056.18          356.70      1543.57         1077.35   
4                1    3902.00         6500.20     11364.40         8663.60   
...            ...        ...             ...          ...             ...   
54634            8    2743.29          519.75      3564.87         2222.55   
54635            0    9709.73         3849.30     14813.24        12095.92   
54636            6   31188.50        13832.00     49184.90        41054.25   
54637            0     165.76        14143.36     15632.71        14280.00   
54638            5    3103.31        14939.19     19711.43        16599.10   

         profit  
0      46242.03  
1        284.47  
2        311.94  
3        466.22  
4       2700.80  
...         ...  
54634   1342.32  
54635   2717.32  
54636   8130.65  
54637   1352.71  
54638   3112.33  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [117]:
# Check your work using the totals above
checking_work = client_data.groupby("order_id")["total price"].sum()
order1 = 2742071
order2 = 2173913
order3 = 6128929
does_it_match1 = checking_work.loc[order1]
does_it_match2 = checking_work.loc[order2]
does_it_match3 = checking_work.loc[order3]
print(round(does_it_match1, 2))
print(round(does_it_match2, 2))
print(round(does_it_match3, 2))

152811.89
162388.71
923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [121]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_clients = client_data.groupby("client_id")["total price"].sum()
top5_clients = top_clients[top_5_clients_df]
top5_clients

client_id
33615     8377308.52
66037    10259514.79
46820     9743794.36
38378    12906550.87
24741    82268892.02
Name: total price, dtype: float64

In [140]:
# Create a summary DataFrame showing the totals for the
# working with Nic on this code
summery_of_data = client_data[client_data["client_id"].isin(top_5_clients_df)]
# for the top 5 clients with the following information:
summery_of_data_df = summery_of_data.groupby("client_id").agg({"qty": "sum", 
                                                               "shipping price": "sum", 
                                                               "total price": "sum", 
                                                               "new total cost": "sum", 
                                                               "profit": "sum", }).reset_index(drop=False)
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summery_of_data_df.sort_values(by="profit", ascending=False)


client_id     qty  shipping price  total price  new total cost       profit
0      24741  239862      5126448.37  82268892.02     45688899.71  36579992.31
2      38378   73667      3429455.40  12906550.87      9634720.98   3271829.89
4      66037   43018      1395151.85  10259514.79      7004482.98   3255031.81
3      46820   75768      1601448.84   9743794.36      7007191.64   2736602.72
1      33615   64313      1828984.89   8377308.52      6175313.91   2201994.61

In [141]:
# Format the data and rename the columns
# to names suitable for presentation.
summery_of_data_df = summery_of_data_df.rename(columns={"client_id": "Client ID", 
                                                        "qty": "Units", 
                                                        "shipping price": "Shipping", 
                                                        "total price": "Total Revenue", 
                                                        "new total cost": "Total Cost", 
                                                        "profit": "Total Profit", 
                                                       })
# Currency should be in millions of dollars. 
# Creating a function to set to millions.
def set_to_million(x):
    return x/1000000

# dividing results by 1,000,000 to get correct units
summery_of_data_df["Shipping"] = summery_of_data_df["Shipping"].apply(set_to_million)
summery_of_data_df["Total Revenue"] = summery_of_data_df["Total Revenue"].apply(set_to_million)
summery_of_data_df["Total Profit"] = summery_of_data_df["Total Profit"].apply(set_to_million)
summery_of_data_df["Total Cost"] = summery_of_data_df["Total Cost"].apply(set_to_million)

# Creating a function for format "$" & "M"
def format_millions(x):
    return "${:.2f}M".format(x)

# Adding format to column results
summery_of_data_df["Shipping"] = summery_of_data_df["Shipping"].apply(format_millions)
summery_of_data_df["Total Revenue"] = summery_of_data_df["Total Revenue"].apply(format_millions)
summery_of_data_df["Total Profit"] = summery_of_data_df["Total Profit"].apply(format_millions)
summery_of_data_df["Total Cost"] = summery_of_data_df["Total Cost"].apply(format_millions)

summery_of_data_df

Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
1      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M
2      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
3      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
4      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M

In [142]:
# Sort the updated data by "Total Profit" form highest to lowest
summery_of_data_df.sort_values(by="Total Profit", ascending=False)

Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
2      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
4      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
3      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
1      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M